src/categorical_algebra/CSets.jl

```diff
@@ -661,31 +657,50 @@ end
 # Sub-C-sets
 ############
 
+""" Sub-C-set of a C-set.
+"""
 const SubCSet{S} = Subobject{<:StructCSet{S}}
 const SubACSet{S} = Subobject{<:StructACSet{S}}
 
 components(A::SubACSet) = map(Subobject, components(hom(A)))
 force(A::SubACSet) = Subobject(force(hom(A)))
 
-""" Construct subobject of C-set from components of inclusion or predicate.
-
-This function constructs a subobject from the components of the inclusion
-morphism, given as a named tuple or as keyword arguments. The components can
-also be specified as predicates (boolean vectors).
+""" Sub-C-set represented componentwise as a collection of subsets.
 """
-function Subobject(X::T, components) where T <: ACSet
-  U = T()
-  components = map(coerce_subob_component, components)
-  copy_parts!(U, X, components)
-  Subobject(ACSetTransformation(components, U, X))
+@auto_hash_equals struct SubACSetComponentwise{
+    Ob<:ACSet, Comp<:NamedTuple} <: Subobject{Ob}
+  ob::Ob
+  components::Comp
+
+  function SubACSetComponentwise(X::Ob, components::NamedTuple) where Ob<:ACSet
+    sets = fin_sets(X)
+    @assert keys(components) ⊆ keys(sets)
+    coerced_components = NamedTuple{keys(sets)}(
+      coerce_subob_component(set, get(components, ob) do; 1:0 end)
+      for (ob, set) in pairs(sets))
+    new{Ob,typeof(coerced_components)}(X, coerced_components)
+  end
 end
+
+Subobject(X::ACSet, components::NamedTuple) =
+  SubACSetComponentwise(X, components)
 Subobject(X::ACSet; components...) = Subobject(X, (; components...))
 
-coerce_subob_component(f::FinFunction{Int}) = collect(f)
-coerce_subob_component(A::SubFinSet{Int}) = collect(hom(A))
-coerce_subob_component(f::AbstractVector{Int}) = f
-coerce_subob_component(pred::Union{AbstractVector{Bool},BitVector}) =
-  findall(pred)
+coerce_subob_component(X::FinSet, subset::SubFinSet) =
```

Comment:
Probably worth documenting whether X is supposed to be the domain or codomain in these error messages. By that I mean, are we asserting that `X ⊆ codom(f)` or `dom(f) ⊆ X`? I was confused by it when reading and then referred to the Subobjects for finset for clarity and saw this

```
SubFinSet(X, f) = Subobject(FinFunction(f, X))
```
which made me even more confused.